src/compiler/types.ts

```diff
@@ -272,6 +272,7 @@ namespace ts {
         NonNullExpression,
         MetaProperty,
         SyntheticExpression,
+        PrivateIdentifierInInExpression,

```

Comment:
I'm not sure we need a `SyntaxKind` here. We already have `PrivateIdentifier`, and we've allowed non-expression tokens like `super` and `import` in "expression" positions for `super()` and `import()`. We would primarily want to just disallow it in places where it can't normally exist as a grammar check in the checker.

Reply:
I'm happy to try and drop this. I didn't want to be the one to break an invariant, but if that bridge has already been crossed then, as you hint, adding this new SyntaxKind is not adding the value I thought it was.
I'll make a start moving this feature to use a standard `BinaryExpression` and report back on any findings I have.

Reply:
I started a (quick and dirty) spike branch with 1 extra commit: https://github.com/bloomberg/TypeScript/commit/023cf75c25486858e1b7229304539cc8b4a5bec1

The general approach does seem to work. Few things that I haven't been able to get working yet:

- I don't think the flow-nodes are being set up right because narrowing isn't happening any more.
- Need to get my head around the binary trampoline in the emitter, as couldn't get it to work in all cases
- `PrivateIdentifier` extends directly from `Node`, so most type asserts had to go via `Node` to get to and from `PrivateIdentifier` <-> `Expression`

Reply:
If you're debugging the compiler, you can use `ts.Debug.formatControlFlowGraph(node.flowNode)` while in the checker to get an idea of what the control flow graph looks like for a given expression.

> - `PrivateIdentifier` extends directly from `Node`, so most type asserts had to go via `Node` to get to and from `PrivateIdentifier` <-> `Expression`

You can just have `PrivateIdentifier` extend from `PrimaryExpression` just like `Identifier`.

It may not be a *real* expression, but our AST sometimes plays fast and loose compared to the spec. As I mentioned before, we treat `super` as a `SuperExpression` rather than just a `Token`.

Reply:
> You can just have PrivateIdentifier extend from PrimaryExpression just like Identifier.

That’s what I was thinking of doing, so very pleased that route is open.

Reply:
The separate `PrivateIdentifierInInExpression` SyntaxKind has been completely removed now. And `PrivateIdentifier` now extends `PrimaryExpression` and this feature now parses to a `BinaryExpression`. Tests remain as they were and are all passing.

The logic for the parser and checker still forks off and reuses the previous logic when this was a separate expression type.
Perhaps this is OK, perhaps not. I'll assume the latter and start to try and integrate the logic closer to the usual BinaryExpression path so. The difference here is that the LHS shouldn't parse or type-check on it's own, unlike a usual BinaryExpression.